### ISSUE TITLE
Excluding of build-automation-generated files (like by "autotool")?

### ISSUE THREAD

**Issue by jschleus** [AlDanial/cloc#151]
I am the maintainer of the FOSS server https://fossies.org that generates a CLOC analysis report for every of the roughly 4000 contained software packages and want to report a very old problem that was pointed out to me by the author of the "xorriso" package already in June 2014:
```
Further i wonder whether it would be possible to keep the autotools generated
scripts out of the CLOC analysis. Their content stems mainly from
my local installation of autotools, not from project specific work. 
... 
The main coding effort is in the .c and .h files. 
Currently this is quite misrepresented by CLOC (extract):

--------------------------------------------------------------------------------
Language           files     blank   comment      code    scale   3rd gen. equiv
--------------------------------------------------------------------------------
C                    121     17411     15705    114211 x   0.77 =       87942.47
Bourne Shell          15      4821      6039     36881 x   3.81 =      140516.61
C/C++ Header          10      5267     19453      7950 x   1.00 =        7950.00
```
As you see the Fossies CLOC reports use the option "--3" (print third-generation language output). Although that output is to be taken with a pinch of salt the report seems to give a completely wrong result. So I excluded then the autotools generated files (and some other files optionally generated by external tools) and the Fossies output was now much more realistic:
```                                                                                                               
C                    121     17411     15705    114211 x   0.77 =       87942.47
C/C++ Header         100      5267     19453      7950 x   1.00 =        7950.00
Bourne Shell           4        75       114       315 x   3.81 =        1200.15
```
Since just CLOC 1.72 was released I were curious what the actual situation is. Here the original output for the xorriso (v1.4.6) package:
```
C                    121     18181     16177    120306 x   0.77 =       92635.62
Bourne Shell          14      3683      4123     23820 x   3.81 =       90754.20
m4                     8      1955       241     17360 x   1.00 =       17360.00
HTML                   4      1472        27      9700 x   1.90 =       18430.00
C/C++ Header         100      5437     20710      8232 x   1.00 =        8232.00
TeX                    1       751      3231      6097 x   1.50 =        9145.50
Tcl/Tk                 1       902       645      4382 x   4.00 =       17528.00
Bourne Again Shell     8       184       195      1297 x   3.81 =        4941.57
make                   1        39        34       287 x   2.50 =         717.50
C++                    2        10        14        40 x   1.51 =          60.40
--------------------------------------------------------------------------------
SUM:                 260     32614     45397    191521 x   1.36 =      259804.79
```
And here for comparison the Fossies output:
```
C                    121     18181     16177    120306 x   0.77 =       92635.62
HTML                   4      1472        27      9700 x   1.90 =       18430.00
C/C++ Header         100      5437     20710      8232 x   1.00 =        8232.00
Tcl/Tk                 1       902       645      4382 x   4.00 =       17528.00
Bourne Again Shell     8       184       195      1297 x   3.81 =        4941.57
Bourne Shell           5       131       211       645 x   3.81 =        2457.45
C++                    2        10        14        40 x   1.51 =          60.40
--------------------------------------------------------------------------------
SUM:                 241     26317     37979    144602 x   1.00 =      144285.04
```
The difference is less terrible but terrible enough to speak from a falsified impression the original CLOC output is giving.

Here the according CLOC options currently used by Fossies:
```
-exclude-dir="autotools,build-aux" --exclude-list-file=cloc_files.excl
```
with cloc_files.excl
```
acinclude.m4
aclocal.m4
announce-gen
autogen.sh
bootstrap
compile
config.guess
config.h.in
config.rpath
config.status
config.sub
configure
configure.ac
configure.in
depcomp
gendocs.sh
gitlog-to-changelog
git-version-gen
gnupload
gnu-web-doc-update
install-sh
libtool
libtool.m4
link-warning.h
ltmain.sh
lt~obsolete.m4
ltoptions.m4
ltsugar.m4
ltversion.in
ltversion.m4
Makefile.am
Makefile.in
mdate-sh
missing
mkinstalldirs
test-driver
texinfo.tex
update-copyright
useless-if-before-free
vc-list-files
ylwrap
```
Ok, that list contains also non-autotools files and is surely imperfect and should so only give an impression what kind of files a suggested CLOC option like for e.g.
```
--ignore-autotools-generated-files
```
or more generally
```
--ignore-build-automation-generated-files
```
(or similar) could be exclude.
Or even that files could be excluded by default and may be only included optionally.

**Comment by AlDanial** [comment]
My first inclination is to add the files above to cloc's internal ignore list with the exclusion reason that they are autogenerated.  But a new switch may be more prudent.

I'll make the update but it will be a week or two away.

**Comment by jschleus** [comment]
Just a small correction to my exclude list: For e.g. the files acinclude.m4, configure.ac and Makefile.am
may be manually created project-specific input parameter files for autotools and not automatically created. But sorry, I  am not familiar with autotools.

**Comment by AlDanial** [comment]
I added a new option, --no-autogen with commit b25d4ab.  Let me know how it works for you.

**Comment by jschleus** [comment]
Hmm, unfortunately I see in a first test using  --no-autogen  no differences in the output (besides of cloc v 1.73 instead of 1.72).
But only listing the to be excluded files by using the option --no-autogen=list works (but the option --help outputs errorneously the option --no-autoconf=list).

**Comment by AlDanial** [comment]
Thanks for catching the documentation error.
My simple testing suggests ``--no-autogen`` works.  I created an empty directory with a C file and a trivial shell script entitled ``mdate-sh`` which is in the exclusion list you provided.  cloc properly ignored it when ``--no-autogen`` is given:
<pre>
    /tmp/x> ll
    total 8
    -rw------- 1 al al 178 Jan 27 11:26 hello.c
    -rw------- 1 al al  15 Jan 27 11:26 mdate-sh

    /tmp/x> cloc .
           2 text files.
           2 unique files.                          
           0 files ignored.

    github.com/AlDanial/cloc v 1.73  T=0.01 s (200.1 files/s, 1400.7 lines/s)
    -------------------------------------------------------------------------------
    Language                     files          blank        comment           code
    -------------------------------------------------------------------------------
    C                                1              0              7              5
    Bourne Shell                     1              0              0              2
    -------------------------------------------------------------------------------
    SUM:                             2              0              7              7
    -------------------------------------------------------------------------------

    /tmp/x> cloc --no-autogen .
           2 text files.
           1 unique file.                          
           1 file ignored.

    github.com/AlDanial/cloc v 1.73  T=0.01 s (99.8 files/s, 1198.1 lines/s)
    -------------------------------------------------------------------------------
    Language                     files          blank        comment           code
    -------------------------------------------------------------------------------
    C                                1              0              7              5
    -------------------------------------------------------------------------------
</pre>

**Comment by jschleus** [comment]
Ok, that can I reproduce.

After some minutes of  perplexity I found out that I use in production (and also in the mentioned test above) the option --report-file=<file>. Using that option the option --no-autogen seems inefficient or ignored.

Probably related I found in the console output (using --report-file/tmp/cloc_report) the following information:
```
1 error:
Unable to read:
Wrote /tmp/cloc_report
```

**Comment by AlDanial** [comment]
Using ``--no-autogen`` and ``--report-file`` works for me.  Using the same setup as above:
<pre>
    /tmp/x> cloc --report-file=/tmp/out.txt --no-autogen .
           2 text files.
           1 unique file.
           1 file ignored.
    Wrote /tmp/out.txt
    /tmp/x> cat /tmp/out.txt 
    github.com/AlDanial/cloc v 1.73  T=0.01 s (91.8 files/s, 1102.0 lines/s)
    -------------------------------------------------------------------------------
    Language                     files          blank        comment           code
    -------------------------------------------------------------------------------
    C                                1              0              7              5
    -------------------------------------------------------------------------------
</pre>

**Comment by jschleus** [comment]
Ah, I should have shown the complete command line I have used since the difference in the behavior seems caused by the difference in the order of the used options:
```
cloc --report-file=/tmp/out.txt --no-autogen .
```
works but
```
cloc --no-autogen --report-file=/tmp/out.txt .
```
not  (at least for me).

For the exemplarily used xorriso (v1.4.6) package mentioned above I got now (using additionally the "--3" option) a  third-generation language sum of 145736.54 (instead of 259804.79 before without using the new option "--no-autogen").

Great, I think that value is more realistic. Thanks!

**Comment by jschleus** [comment]
A small additional information that may help to find the bug: Just for curiosity I used the "meaningless" option
```
--no-autogen=dummy
```
than it works also independent of the options order but I got the already above mentioned error output:
```
1 error:
Unable to read:  dummy
```
Replacing "dummy" by "." it works even without that error message replacing "dummy" by "-" (stdin?) it hangs (but a CTRL-c let it process also without any error output).

**Comment by AlDanial** [comment]
I messed up command line parsing of ``--no-autogen`` with and without ``list``.  Should be fixed with commit d400736; please test with that one.

**Comment by jschleus** [comment]
Great, now `--no-autogen` works position-indepently.

Two small comments:

1) Just for clarifying: The option `--no-autogen=list` now no longer works but seems replaced by  `--no-autogen list` (that seems better to me but the `cloc --help` output must still be adapted).

2) I did some simple tests by  prepending and appending some strings to the exclusion test filenames (for e.g. mdate-sh) to check if  only the exact filenames of the list are excluded. So far ok but I found that generally all manual page files (with an extension .[0-9] and .man) are ignored (independently of `--no-autogen`).
